src/coreclr/System.Private.CoreLib/src/System/Reflection/Emit/ILGenerator.cs

```diff
@@ -1098,12 +1149,39 @@ public virtual void MarkLabel(Label loc)
                 throw new ArgumentException(SR.Argument_InvalidLabel);
             }
 
-            if (m_labelList[labelIndex] != -1)
+            if (m_labelList[labelIndex].m_pos != -1)
             {
                 throw new ArgumentException(SR.Argument_RedefinedLabel);
             }
 
-            m_labelList[labelIndex] = m_length;
+            m_labelList[labelIndex].m_pos = m_length;
+
+            int depth = m_labelList[labelIndex].m_depth;
+            if (depth < 0)
+            {
+                // Unknown depth for this label, indicating that it hasn't been used yet.
+                // If m_curDepth is unknown, we're in the Backward branch constraint case. See ECMA-335 III.1.7.5.
+                // The m_depthAdjustment field will compensate for violations of this constraint, as we
+                // discover them. That is, here we assume a depth of zero. If a (later) branch to this label
+                // has a positive stack depth, we'll record that as the new depth and add the delta into
+                // m_depthAdjustment.
```

Comment:
Do we need to handle the violation or can we just assume depth of zero? What does the existing code do?

ECMA here states "it would clearly be invalid CIL if a later branch instruction to X were to have a non-empty evaluation stack"

Reply:
@jkotas should probably chime in here. In our discussions in January about this Jan indicated that the ECMA rule is not enforced at all and there are known violations of it that are supported today. So this code is written to support IL that violates the rule.

Reply:
I have submitted https://github.com/dotnet/runtime/pull/70760 to document the fact that the backward branch rule is not a thing.

Reply:
Based on that, I assume we should keep the code above.